### PR TITLE
[BUG/MAINT] Fixed factor-vae and added GPU / Cleaned code 

### PR DESCRIPTION
**Pull request by linesd**
- Fixed issues in factor-vae regarding incorrect use of latent_dist rather than latent_sample
- Cleaned up the implementation
- Added GPU functionality for factor-vae

### CODE REVIEW

File: disvae/losses.py [hunk 1]
```diff
@@ -204,27 +212,27 @@ def __call__(self, data, model, optimizer, is_train, storer):
         batch_size = data.size(dim=0)
         half_batch_size = batch_size // 2
         data = data.split(half_batch_size)
-        data1 = data[0].to(self.device)
-        data2 = data[1].to(self.device)
+        data1 = data[0]
+        data2 = data[1]
```
Comment (alecokas) [inline disvae/losses.py]: Maybe it is better to name ```data1``` and ```data2``` something more descriptive?

Comment (linesd) [reply disvae/losses.py]: Happy to do so, but its what to call them. `data1` trains the vae and the discriminator. `data2` just trains the discriminator. Any suggestions?

Comment (linesd) [inline disvae/losses.py]: Quick opinion @YannDubs and @alecokas. Currently, `data1` and `data2` will be the same every epoch. Wondering if I should switch them around each epoch?

Comment (alecokas) [reply disvae/losses.py]: OK that is fine if they're officially called that in the paper. :+1:

Comment (alecokas) [reply disvae/losses.py]: I'm not sure. Let me get back to you.

File: disvae/losses.py [hunk 2]
```diff
@@ -190,12 +190,20 @@ class FactorKLoss(BaseLoss):
...
+                 disc_kwargs=dict(neg_slope=0.2, latent_dim=10, hidden_units=1000),
+                 optim_kwargs=dict(lr=1e-4, betas=(0.5, 0.9))):
         super().__init__(is_color)
         self.beta = beta
         self.device = device
-        self.discriminator = discriminator
-        self.optimizer_d = optimizer_d
+
+        self.discriminator = Discriminator(neg_slope=disc_kwargs['neg_slope'],
+                                           latent_dim=disc_kwargs['latent_dim'],
```
Comment (YannDubs) [inline disvae/losses.py]: use `Discriminator(**disc_kwargs)`, i.e. just redirect the arguments, such that you can change all arguments of `Discriminator` from the loss (without having to manually redirect as you are doing)

Same for optimizer

File: disvae/losses.py (hunk 1 above)
Comment (YannDubs) [reply disvae/losses.py]: I don't think you need to: the data is being shuffled during training in any case

File: disvae/vae.py [hunk 3]
```diff
@@ -50,21 +50,28 @@ def reparameterize(self, mean, logvar):
...
         """
         latent_dist = self.encoder(x)
         latent_sample = self.reparameterize(*latent_dist)
-        reconstruct = self.decoder(latent_sample)
-        if self.is_color:
-            reconstruct = reconstruct * 255
-        return reconstruct, latent_dist
+
+        if get_sample:
+            return latent_sample
```
Comment (YannDubs) [inline disvae/vae.py]: not a big fan of changing the number of outputs of a function using a bool. If it were me i would define a simple helper method in `VAE`:

```
def sample_latent(self, x):
        latent_dist = self.encoder(x)  
        latent_sample = self.reparameterize(*latent_dist)
        return latent_sample
```

I think it's clear from the code what you are doing

File: disvae/losses.py [hunk 4]
```diff
@@ -204,27 +212,27 @@ def __call__(self, data, model, optimizer, is_train, storer):
...
         kl_loss = _kl_normal_loss(*latent_dist, storer)
 
-        # Run latent distribution through discriminator
-        d_z = self.discriminator(latent_dist)
+        # Run latent sample through discriminator
+        d_z = self.discriminator(latent_sample)
 
         # Calculate the total correlation (TC) loss term
-        tc_loss = (d_z[:, :, :1] - d_z[:, :, 1:]).mean()
+        tc_loss = (d_z[:, :1] - d_z[:, 1:]).mean()
```
Comment (YannDubs) [inline disvae/losses.py]: so decided to keep 2 outputs? Maybe for interest of time we can change/test that after the poster

File: disvae/losses.py [hunk 5]
```diff
@@ -325,30 +332,29 @@ def _kl_normal_loss(mean, logvar, storer=None):
     return total_kl
 
 
-def _permute_dims(latent_dist):
+def _permute_dims(latent_sample):
```
Comment (YannDubs) [inline disvae/losses.py]: how are the results compared to before?

**Comment by YannDubs** [comment]
small comments: feel free to merge once you address them

File: disvae/losses.py (hunk 4 above)
Comment (linesd) [reply disvae/losses.py]: I made a comment on the original pull request but I don't think you saw it, so repeated here:

I took another look at the paper and they say in Appendix A:

> We use the same 6 layer MLP discriminator with 1000 hidden units per layer and leaky ReLU (lReLU) non-linearity, that outputs 2 logits in all FactorVAE experiments.

So I think the discriminator implementation is actually correct. Is the constraint enforced by this line of code?:

`d_tc_loss = 0.5 * (F.cross_entropy(d_z, zeros) + F.cross_entropy(d_z_perm, ones))`

File: disvae/vae.py (hunk 3 above)
Comment (linesd) [reply disvae/vae.py]: Agree, I'll do this.

File: disvae/losses.py (hunk 2 above)
Comment (linesd) [reply disvae/losses.py]: I tried this and had some issues when passing an empty dictionary. Let me give it another go and see if I can find what the issue was.

File: disvae/losses.py (hunk 5 above)
Comment (linesd) [reply disvae/losses.py]: I'm currently training on the server (300 epochs), I'll post the results in slack when finished (+- couple hours)

File: disvae/losses.py (hunk 2 above)
Comment (linesd) [reply disvae/losses.py]: Not entirely sure what I was doing (clearly something a bit stupid), but is now working.

Comment (linesd) [reply disvae/losses.py]: Have changed ...

File: disvae/losses.py (hunk 4 above)
Comment (linesd) [reply disvae/losses.py]: Agreed, we can explore this more after the poster.
